crimson: add support for backfill, part 0

rzarzynski opened the pull request:
This is initial PR of the backfill-for-crimson effort. Part no. 1 part will bring unit tests for the backfill FSM introduced here.

## TODO
- [x] move `BackfillState` from `PG` to `PGRecovery`,
- [x] drop the DNM commit for `PeeringState`,
- [ ] glue with regular IO blocking when it's ready,
- [ ] `BackfillListener::enqueue_drop` with the corresponding flush,
- [ ] `BackfillListener::budget_available()`.

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

tchaikov commented:
@rzarzynski could you connect this PR with one of the backfill cards in https://trello.com/b/lbHkWKxh/crimson ?

rzarzynski commented:
![](https://github.trello.services/images/mini-trello-icon.png) [backfill: advance through object space](https://trello.com/c/AUKPvn0c/10-backfill-advance-through-object-space)

rzarzynski commented:
![](https://github.trello.services/images/mini-trello-icon.png) [backfill: scan on-disk objects across replicas](https://trello.com/c/WaUH1pZu/1-backfill-scan-on-disk-objects-across-replicas)

rzarzynski commented on src/crimson/osd/pg.cc:
```diff
@@ -865,4 +865,50 @@ void PG::handle_rep_op_reply(crimson::net::Connection* conn,
   backend->got_rep_op_reply(m);
 }
 
+void PG::request_replica_scan(
```
This is an artifact from the past. Backfill has been forked form the recovery branch when it wasn't using `PGRecovery` yet. Need to move it out.

athanatos commented on src/crimson/osd/backfill_state.h:
```diff
@@ -29,61 +32,191 @@ struct BackfillState {
 
-  struct Flushed : sc::event<Flushed> {
+  struct ObjectPushed : sc::event<ObjectPushed> {
+    // TODO: implement replica management; I don't want to follow
```
Ah, you mean that backfill doesn't care how many replicas there are since "Push to replicas" is a foundational operation.  That makes sense to me.

athanatos commented on src/osd/PeeringState.h:
```diff
@@ -245,7 +245,91 @@ struct PeeringCtxWrapper {
+ * 1) begin == end == hobject_t() indicates the the interval is unpopulated
+ * 2) Else, objects contains all objects in [begin, end)
+ */
+struct BackfillInterval {
```
I'd like to stop adding things to PeeringState.h.  If possible, can we put this in another file?  Ideally with the backfill machinery?

rzarzynski replied:
How about `peering_state_types.h` per analogy to `osd_types.h` and `bluestore_types.h`?

athanatos replied:
Yeah, that would be fine.

rzarzynski replied:
Moved to `recovery_types.h`.

athanatos commented on src/osd/PeeringState.cc:
```diff
@@ -6453,6 +6466,7 @@ boost::statechart::result PeeringState::GetInfo::react(const QueryState& q)
 void PeeringState::GetInfo::exit()
 {
   context< PeeringMachine >().log_exit(state_name, enter_time);
+  ceph_assert(peer_info_requested.empty());
```
Is this actually true in general?  What if we get an interval change before GetInfo completes?

rzarzynski replied:
After rereading I see a path leading to assertion failure. If `prior_set.pg_down` is set, the `probe` container **can still** enlist shards that primary should to send `pg_query_t::INFO` to. Therefore, they can be present in the `peer_info_requested` when `exit()` is called in the effect of dispatching `IsDown` from the constructor.

When `GetInfo` handles  `MNotifyRec` that affects `last_epoch_started`, `peer_info_requested` is cleared from the OSDs that aren't in the probe set anymore.

How about:

```cpp
  ceph_assert(peer_info_requested.empty() || prior_set.pg_down);
```
?

athanatos replied:
I still don't think that's right.  If Started (parent of GetInfo) gets an AdvMap event for which should_restart_peering is true, it will transition directly to Reset.  The inner states (Primary/Peering/GetInfo) here would each have ::exit() and then their destructor called.  In that case, peer_info_requested may well be non-empty, and that's perfectly ok.  peer_info_requested is a member of GetInfo, and will cease to exist.  When we next enter GetInfo, it'll be a new and empty map.

rzarzynski replied:
Thanks! I will drop the commit.

athanatos commented on src/osd/PeeringState.cc:
```diff
@@ -1461,7 +1472,6 @@ map<pg_shard_t, pg_info_t>::const_iterator PeeringState::find_best_info(
   /* See doc/dev/osd_internals/last_epoch_started.rst before attempting
    * to make changes to this process.  Also, make sure to update it
    * when you find bugs! */
-  eversion_t min_last_update_acceptable = eversion_t::max();
```
Why?  Is this just to delay declaration?

athanatos replied:
Bluntly, this code gets looked at every time there's an emergency with peering.  If you are going to make minor cosmetic changes, you need to be more specific in the commit message as to why so that future readers can comfortably skip your commit.

rzarzynski replied:
Yes, the sole purpose of the change was to move the declaration closer to its usage. Commented that in the commit description.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -13206,9 +13208,6 @@ uint64_t PrimaryLogPG::recover_backfill(
     }
   }
 
-  hobject_t backfill_pos =
```
Same here, explain why you are doing this for future git-blame readers.

rzarzynski replied:
The shuffling was only to move closer to the usage. Added a note to the commit message saying this is intended to be a refactoring only.

athanatos commented on src/crimson/osd/backfill_state.cc:
```diff
@@ -0,0 +1,523 @@
+  bs().pending_backfill_updates.clear();
+  if (Enqueuing::all_enqueued(ps(),
+                              bs().backfill_info,
+                              bs().peer_backfill_info)) {
```
Just for clarity, if all_enqueued() here, then we've actually completed our pushes, right (otherwise we'd be in Waiting).  Can we actually get here under those circumstances?

rzarzynski replied:
This is executed only once – after the initialization of entire FSM. The `peer_backfill_info` map, which `all_enqueued()` depends on, is updated with `pg_info_t::last_backfill` and that's why I'm little bit paranoiac here.

athanatos commented on src/crimson/osd/backfill_state.cc:
```diff
@@ -0,0 +1,523 @@
+  const BackfillInterval& backfill_info) const
+{
+  return backfill_info.begin <= earliest_peer_backfill(peer_backfill_info) &&
+	 backfill_info.extends_to_end() == false;
```
!backfill_info.extends_to_end()

athanatos commented on src/crimson/osd/backfill_state.cc:
```diff
@@ -0,0 +1,523 @@
+{
+  return replica_backfill_info.empty() && \
+         replica_backfill_info.begin <= local_backfill_info.begin && \
+         replica_backfill_info.extends_to_end() == false;
```
!replica_backfill_info.extends_to_end()

athanatos commented on src/crimson/osd/backfill_state.h:
```diff
@@ -0,0 +1,323 @@
+    // to Enqueuing will happen.
+    sc::result react(ObjectPushed);
+    sc::result react(ReplicaScanned);
+
```
These static predicates need interface comments in this file explaining what they mean and how they fit into the larger invariants.

rzarzynski replied:
Commented.

athanatos commented on src/crimson/osd/backfill_state.cc:
```diff
@@ -0,0 +1,523 @@
+
+  while (!bs().backfill_info.empty()) {
+    if (!ls().budget_available()) {
+      post_event(RequestWaiting{});
```
Why not make enqueue_push() simply wait until some of the in-progress ones complete.  Seems like that would prevent this statemachine from needing to know about it.

rzarzynski replied:
IIUC *wait until some of the in-progress ones complete* would be blocking in that case, so some threading, green or OS-level, would be required. With the `budget_available()` I wanted to avoid dependency on `seastar::thread`.

athanatos replied:
Oh, I missed that.  Yeah, that's fine.

athanatos commented on src/crimson/osd/backfill_state.h:
```diff
@@ -0,0 +1,323 @@
+    bool should_rescan_primary(
+      const std::map<pg_shard_t, BackfillInterval>& peer_backfill_info,
+      const BackfillInterval& backfill_info) const;
+
```
This bit here would greatly benefit from some comments.  A comment explaining what loop_advancer_t does to the state invariant would help massively for someone reading this code.  A better name would also help.

In a wider sense, I'm not really sure I get why loop_advancer_t::advance isn't just a helper function invoked at the end of remove_on_peers and update_on_peers.  It's really not correct to not call advance() on the returned value, right?

athanatos replied:
How about trim_backfilled_object_from_intervals?

rzarzynski replied:
Right, will rework.

rzarzynski replied:
Commented.

athanatos commented on src/crimson/osd/recovery_backend.cc:
```diff
@@ -42,3 +45,227 @@ void RecoveryBackend::clean_up(ceph::os::Transaction& t,
+    spg_t(pg.get_pgid().pgid, pg.get_primary().shard));
+  reply->set_priority(pg.get_recovery_op_priority());
+  (void)m.get_connection()->send(std::move(reply));
+  // XXX: whh not PeeringState::RecoveryDone?
```
Probably because you don't want to hold up the message processing on the async peering state event.

athanatos commented on src/crimson/osd/recovery_backend.cc:
```diff
@@ -42,3 +45,227 @@ void RecoveryBackend::clean_up(ceph::os::Transaction& t,
+      return backend->list_objects(start, limit).then(
+        [this, &start, &version_map] (auto ret) {
+          auto& [objects, next] = ret;
+          return seastar::parallel_for_each(
```
This may not actually be a good idea as it might saturate the IO backend with non-urgent work.  Once we get further along, we may choose to allow some parallelism here, but for now now I'd rather just make it linear.

rzarzynski replied:
Sure. TODO.

athanatos commented:
The approach here seems reasonable as far as I can tell.  backfill_state.h needs short, but clear comments for any important predicates to allow readers to quickly understand their usage and purpose.

I'll circle back once the first round of comments has been addressed.

athanatos commented:
It seems plausible to me, btw, that BackfillState could be backported to classic.  That's the intent, right?

rzarzynski commented:
> It seems plausible to me, btw, that BackfillState could be backported to classic. That's the intent, right?

Yes, retrofitability was the goal. Though, the next implementation of `BackfillListener` & co. will be one in the unit test. After some (rather long) time seasoned in crimson with the help of teuthology, we could start thinking about backporting.

athanatos commented:
Dunno, I think once you have solid unit test coverage there isn't any real reason to wait.

rzarzynski commented:
Addressing the comments in progress.
Pushed updated version with a lot of cleanups and code shuffling (moved `BackfillState` into `PGRecovery`). Additionally, https://github.com/ceph/ceph/pull/35130/commits/2060dd3bc2f4fb1e796f769bb26ad68e027554af addresses the synchronization between peering and backfill.

rzarzynski commented on src/crimson/osd/backfill_facades.h:
```diff
@@ -35,6 +35,10 @@ struct BackfillState::PeeringFacade {
                                              const pg_stat_t &stats) {
     return peering_state.update_complete_backfill_object_stats(hoid, stats);
   }
+
```
Update the commit title. It should be `crimson/osd: settle BackfillState in PGRecovery`.

rzarzynski commented on src/crimson/osd/backfill_state.h:
```diff
@@ -29,61 +32,191 @@ struct BackfillState {
+public:
+
   // states
-  struct Crashed : sc::state<Crashed, BackfillMachine>, NamedState {
```
Hmm, this could be implemented using a class with debugs in both ctor and dtor + a macro to avoid RTTI and the potentially ugly outcome of `typeid(T).name()`.

rzarzynski replied:
Refactored with [`boost::typeindex::type_id<T>().pretty_name()`](https://www.boost.org/doc/libs/1_66_0/doc/html/boost_typeindex/getting_started.html).

tchaikov commented on src/crimson/osd/shard_services.cc:
```diff
@@ -283,4 +287,34 @@ HeartbeatStampsRef ShardServices::get_hb_stamps(int peer)
   return stamps->second;
 }
 
+seastar::future<> ShardServices::send_alive(const epoch_t want)
```
could you update the commit message noting what "rework" stands for?

tchaikov commented on src/crimson/osd/shard_services.h:
```diff
@@ -43,6 +43,7 @@ namespace crimson::osd {
 class ShardServices : public md_config_obs_t {
   using cached_map_t = boost::local_shared_ptr<const OSDMap>;
   OSDMapService &osdmap_service;
+  const int &whoami;
```
nit, pass by value would suffice and probably more efficient.

tchaikov commented on src/crimson/osd/backfill_state.h:
```diff
@@ -0,0 +1,119 @@
+#include <boost/statechart/state.hpp>
+#include <boost/statechart/state_machine.hpp>
+#include <boost/statechart/transition.hpp>
+#include <boost/statechart/event_base.hpp>
```
nit, order the `#include` in the alphabetical order.

tchaikov commented on src/crimson/osd/backfill_state.h:
```diff
@@ -0,0 +1,119 @@
+
+namespace sc = boost::statechart;
+
+struct BackfillState {
```
just out of curiosity, why you didn't make it a namespace? as this struct does not hold any member variables.

rzarzynski replied:
Later commits will bring the members:

```diff
+private:
+  hobject_t last_backfill_started;
+  BackfillInterval backfill_info;
+  std::map<pg_shard_t, BackfillInterval> peer_backfill_info;
+  BackfillMachine backfill_machine;
+  std::unique_ptr<ProgressTracker> progress_tracker;
```

tchaikov commented on src/crimson/osd/backfill_state.cc:
```diff
@@ -455,6 +455,11 @@ BackfillState::Done::Done(my_context ctx)
+// -- Crashed
+BackfillState::Crashed::Crashed()
+{
+  ceph_assert("{}: this should not happen" == nullptr);
```
nit, might want to use `ceph_abort(...)` in this case.

rzarzynski replied:
Switched other occurrences as well.

tchaikov commented on src/crimson/osd/pg_recovery.cc:
```diff
@@ -453,7 +453,17 @@ void PGRecovery::enqueue_push(
+                 __func__, target, obj, v);
+  std::ignore = pg->get_recovery_backend()->recover_object(obj, v).\
+  handle_exception([] (auto) {
+    ceph_assert("got exception on backfill's push" == nullptr);
```
```suggestion
    ceph_abort("got exception on backfill's push");
```

rzarzynski commented on src/crimson/osd/backfill_state.cc:
```diff
@@ -0,0 +1,448 @@
+{
+  std::for_each(std::begin(result.pbi_targets), std::end(result.pbi_targets),
+    [this, &peer_backfill_info] (const auto& bt) {
+      peer_backfill_info.at(bt).pop_front();
```
I think `at()` was chosen because it offers:
* more concise code than the `find()`-based counterpart (assumed here that shorter code is easier to read),
* the `const` restriction. In contrast to `operator[]`, `at()` has a `const`-qualified variant.

rzarzynski commented on src/crimson/osd/backfill_state.h:
```diff
@@ -257,4 +258,35 @@ struct BackfillState::BackfillListener {
+  BackfillMachine& backfill_machine;
+  std::map<hobject_t, op_source_t> registry;
+
+  BackfillState& bs() {
```
Yeah, will try to find something more descriptive but still short.

rzarzynski replied:
Appended https://github.com/ceph/ceph/pull/35130/commits/d80aaca1716459e78fe77155e7f2d0ddd32d8f87 to tackle this. `peering_state()`, `backfill_state()`, `backfill_listener()` are used instead. To preserve formatting in some places aliasing has been introduced.

tchaikov commented:
```
15:22:54 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 0.136055/0.15826  => rejected
15:22:54 - INFO     - cbt      - prefill/incerta08.front.sepia.ceph.com/0: cpu_cycles_per_op: (or (less) (near 0.05)):: 1015454.4017857143/1413289.3359375  => accepted
15:22:54 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 34.0/40.0  => rejected
15:22:54 - INFO     - cbt      - prefill/incerta08.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 2.00)):: 15.5242/9.2376  => accepted
15:22:54 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.457785/0.394879  => rejected
15:22:54 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 0.136046/0.150794  => rejected
15:22:54 - INFO     - cbt      - prefill/incerta08.front.sepia.ceph.com/1: cpu_cycles_per_op: (or (less) (near 0.05)):: 1015454.4017857143/1402333.6046511629  => accepted
15:22:54 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 34.0/38.0  => rejected
15:22:54 - INFO     - cbt      - prefill/incerta08.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 2.00)):: 16.0/8.96289  => accepted
15:22:54 - WARNING  - cbt      - prefill/incerta08.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.459307/0.39322  => rejected
15:22:54 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 36.1015/38.7624  => rejected
15:22:54 - INFO     - cbt      - rand/incerta08.front.sepia.ceph.com/0: cpu_cycles_per_op: (or (less) (near 0.05)):: 278951.13404553063/283522.5054280761  => accepted
15:22:54 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 9241.0/9923.0  => rejected
15:22:54 - INFO     - cbt      - rand/incerta08.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 2.00)):: 15470.5/16456.2  => accepted
15:22:54 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.00172719/0.00160823  => rejected
15:22:54 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 36.9/40.9819  => rejected
15:22:54 - INFO     - cbt      - rand/incerta08.front.sepia.ceph.com/1: cpu_cycles_per_op: (or (less) (near 0.05)):: 272916.2731523139/270700.9530409081  => accepted
15:22:54 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 9446.0/10491.0  => rejected
15:22:54 - INFO     - cbt      - rand/incerta08.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 2.00)):: 17803.5/15511.2  => accepted
15:22:54 - WARNING  - cbt      - rand/incerta08.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.00169007/0.00152086  => rejected
15:22:54 - WARNING  - cbt      - 12 tests failed out of 20
```

tchaikov commented:
jenkins test crimson perf

tchaikov commented:
jenkins test make check

rzarzynski commented on src/crimson/osd/backfill_state.cc:
```diff
@@ -0,0 +1,543 @@
+  // the machine to the state.
+  //ceph_assert(ps().is_backfill_target(evt.from));
+  if (waiting_on_backfill.erase(evt.from)) {
+    bs().peer_backfill_info.emplace(evt.from, std::move(evt.result));
```
`bs().peer_backfill_info[evt.from] = std::move(evt.result)` instead. Entries for all targets are already added by `Initial::Initial`.

rzarzynski commented on src/crimson/osd/backfill_state.cc:
```diff
@@ -0,0 +1,543 @@
+boost::statechart::result
+BackfillState::ReplicasScanning::react(ReplicaScanned evt)
+{
+  // TODO: maybe we'll be able to move waiting_on_backfill from
```
This `TODO` is already handled. Remove.

rzarzynski commented on src/crimson/osd/backfill_state.cc:
```diff
@@ -0,0 +1,543 @@
+{
+  // TODO: maybe we'll be able to move waiting_on_backfill from
+  // the machine to the state.
+  //ceph_assert(ps().is_backfill_target(evt.from));
```
Uncomment or drop.

rzarzynski commented on src/crimson/osd/recovery_backend.cc:
```diff
@@ -159,7 +159,23 @@ seastar::future<> RecoveryBackend::handle_scan_digest(
                  __func__);
-  ceph_assert("Not implemented" == nullptr);
+  // Check that from is in backfill_targets vector
+  //ceph_assert(pg.get_peering_state().is_backfill_target(m.from));
```
Yup, uncommented.

tchaikov commented on src/crimson/osd/backfill_facades.h:
```diff
@@ -42,6 +42,10 @@ struct BackfillState::PeeringFacade {
     return peering_state.update_complete_backfill_object_stats(hoid, stats);
   }
 
+  decltype(auto) is_backfilling() const {
```
nit, might want to s/decltype(auto)/bool/ for better readability? i understand that `PeeringFacade` is proxying the `is_backfilling()` calls to underlying `peering_state`, and`decltype(auto)` conveys the idea of "forwarding the call" perfectly. but i think the type is also part of the documentation, probably `bool` helps the reader to get the point more quickly?

tchaikov commented on src/crimson/osd/recovery_backend.cc:
```diff
@@ -60,6 +60,82 @@ void RecoveryBackend::WaitForObjectRecovery::stop() {
+void RecoveryBackend::handle_backfill_finish(
+  MOSDPGBackfill& m)
+{
+  logger().debug("{}",
```
nit, i'd suggest avoid wrapping lines proactively.

rzarzynski replied:
Changed in entire `recovery_backend.cc`.

rzarzynski commented:
```
15:09:54 - WARNING  - cbt      - prefill/incerta03.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 16.2583/29.3698  => rejected
15:09:54 - INFO     - cbt      - prefill/incerta03.front.sepia.ceph.com/0: cpu_cycles_per_op: (or (less) (near 0.05)):: 0.0/0.0  => accepted
15:09:54 - WARNING  - cbt      - prefill/incerta03.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 4162.0/7518.0  => rejected
15:09:54 - INFO     - cbt      - prefill/incerta03.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 2.00)):: 104.987/367.022  => accepted
15:09:54 - WARNING  - cbt      - prefill/incerta03.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.00381061/0.00210483  => rejected
15:09:54 - WARNING  - cbt      - prefill/incerta03.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 16.2005/29.5615  => rejected
15:09:54 - INFO     - cbt      - prefill/incerta03.front.sepia.ceph.com/1: cpu_cycles_per_op: (or (less) (near 0.05)):: 0.0/0.0  => accepted
15:09:54 - WARNING  - cbt      - prefill/incerta03.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 4147.0/7567.0  => rejected
15:09:54 - INFO     - cbt      - prefill/incerta03.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 2.00)):: 84.1606/360.056  => accepted
15:09:54 - WARNING  - cbt      - prefill/incerta03.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.00381657/0.00209459  => rejected
15:09:54 - WARNING  - cbt      - rand/incerta03.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 112.625/244.842  => rejected
15:09:54 - INFO     - cbt      - rand/incerta03.front.sepia.ceph.com/0: cpu_cycles_per_op: (or (less) (near 0.05)):: 0.0/0.0  => accepted
15:09:54 - WARNING  - cbt      - rand/incerta03.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 28832.0/62679.0  => rejected
15:09:54 - INFO     - cbt      - rand/incerta03.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 2.00)):: 1269.61/1878.23  => accepted
15:09:54 - WARNING  - cbt      - rand/incerta03.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.000551746/0.000252231  => rejected
15:09:54 - WARNING  - cbt      - rand/incerta03.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 111.72/248.001  => rejected
15:09:54 - INFO     - cbt      - rand/incerta03.front.sepia.ceph.com/1: cpu_cycles_per_op: (or (less) (near 0.05)):: 0.0/0.0  => accepted
15:09:54 - WARNING  - cbt      - rand/incerta03.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 28600.0/63488.0  => rejected
15:09:54 - INFO     - cbt      - rand/incerta03.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 2.00)):: 1432.15/2090.52  => accepted
15:09:54 - WARNING  - cbt      - rand/incerta03.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.000556166/0.000249042  => rejected
```

rzarzynski commented:
jenkins retest this please

tchaikov commented:
jenkins test crimson perf

rzarzynski commented:
jenkins test crimson perf

rzarzynski commented:
jenkins test crimson perf

rzarzynski commented:
Repeated the perf check manually on `incerta07`:

```
(venv) [rzarzynski@incerta07 cbt]$ python /work/cbt/compare.py -b /tmp/wip-crimson-backfill-ref/ -a /tmp/wip-crimson-backfill -v
03:50:14 - DEBUG    - cbt      - Nodes : incerta07.front.sepia.ceph.com
03:50:14 - DEBUG    - cbt      - Nodes : incerta07.front.sepia.ceph.com
03:50:14 - DEBUG    - cbt      - Nodes : incerta07.front.sepia.ceph.com
03:50:14 - INFO     - cbt      - prefill/incerta07.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 17.9427/18.0478  => accepted
03:50:14 - INFO     - cbt      - prefill/incerta07.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 4593.0/4620.0  => accepted
03:50:14 - INFO     - cbt      - prefill/incerta07.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 2.00)):: 330.347/614.208  => accepted
03:50:14 - INFO     - cbt      - prefill/incerta07.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.00347031/0.00344579  => accepted
03:50:14 - INFO     - cbt      - prefill/incerta07.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 18.1544/18.0327  => accepted
03:50:14 - INFO     - cbt      - prefill/incerta07.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 4647.0/4616.0  => accepted
03:50:14 - INFO     - cbt      - prefill/incerta07.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 2.00)):: 331.989/617.077  => accepted
03:50:14 - INFO     - cbt      - prefill/incerta07.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.00343893/0.00344441  => accepted
03:50:14 - INFO     - cbt      - rand/incerta07.front.sepia.ceph.com/0: bandwidth: (or (greater) (near 0.05)):: 163.422/167.33  => accepted
03:50:14 - INFO     - cbt      - rand/incerta07.front.sepia.ceph.com/0: iops_avg: (or (greater) (near 0.05)):: 41836.0/42836.0  => accepted
03:50:14 - INFO     - cbt      - rand/incerta07.front.sepia.ceph.com/0: iops_stddev: (or (less) (near 2.00)):: 1346.2/1875.52  => accepted
03:50:14 - INFO     - cbt      - rand/incerta07.front.sepia.ceph.com/0: latency_avg: (or (less) (near 0.05)):: 0.000376841/0.000367896  => accepted
03:50:14 - INFO     - cbt      - rand/incerta07.front.sepia.ceph.com/1: bandwidth: (or (greater) (near 0.05)):: 164.57/168.476  => accepted
03:50:14 - INFO     - cbt      - rand/incerta07.front.sepia.ceph.com/1: iops_avg: (or (greater) (near 0.05)):: 42129.0/43129.0  => accepted
03:50:14 - INFO     - cbt      - rand/incerta07.front.sepia.ceph.com/1: iops_stddev: (or (less) (near 2.00)):: 1241.57/1834.99  => accepted
03:50:14 - INFO     - cbt      - rand/incerta07.front.sepia.ceph.com/1: latency_avg: (or (less) (near 0.05)):: 0.000374262/0.000365504  => accepted
03:50:14 - INFO     - cbt      - All 16 tests passed.
```
